### PR TITLE
Parallel coordinates

### LINKED ISSUE #85: Make use of parameter-parallel-coordinates from webviz subsurface
https://webviz-subsurface-example.azurewebsites.net/parameter-parallel-coordinates

### PR DESCRIPTION
**Pull request by xjules**
Resolves #85

- Ensemble gets a new method for exporting all parameters as a single dataframe
- new view "Parameter Comparison view"
- make use of `webvize-core-components` `wcc.Select`

### CODE REVIEW

**Comment by xjules** [comment]
Screenshot:
image

File: ertviz/controllers/parameter_comparison_controller.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        colors = []
+        for idx, (ensemble_id, color) in enumerate(selected_ensembles.items()):
+            ensemble = load_ensemble(parent, ensemble_id)
+            ens_key = str(ensemble)
+            df = ensemble.parameters_df(parameters)
+            df["ensemble_id"] = idx
+            data[ens_key] = df.copy()
+            colors.append([idx / len(selected_ensembles), color["color"]])
+            colors.append([(idx + 1) / len(selected_ensembles), color["color"]])
+        parent.parallel_plot = ParalellCoordinates(data, colors)
```
Comment (oysteoh) [inline ertviz/controllers/parameter_comparison_controller.py:34]: ParalellCoordinates -> ParallelCoordinates

File: ertviz/controllers/parameter_comparison_controller.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+from ertviz.models import (
+    load_ensemble,
+    ParalellCoordinates,
+)
+
+
+def parameter_comparison_controller(parent, app):
+    @app.callback(
+        Output(
+            {"id": parent.uuid("paralell-coor"), "type": parent.uuid("graph")},
```
Comment (oysteoh) [inline ertviz/controllers/parameter_comparison_controller.py:12]: paralell-coor -> parallel-coor

File: ertviz/plugins/_parameter_comparison.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+                html.Div(
+                    id=self.uuid("ensemble-content"),
+                    children=ensemble_selector_view(parent=self),
+                ),
+                html.Div(
+                    id=self.uuid("paralell-coor-content"),
+                    children=[
+                        html.H5("Multi parameter selector:"),
+                        parameter_selector_view(parent=self),
+                        paralell_coordinates_view(parent=self),
```
Comment (oysteoh) [inline ertviz/plugins/_parameter_comparison.py:36]: paralell_coordinates_view -> parallel_coordinates_view

File: ertviz/plugins/_parameter_comparison.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    @property
+    def layout(self):
+        return html.Div(
+            [
+                html.Div(
+                    id=self.uuid("ensemble-content"),
+                    children=ensemble_selector_view(parent=self),
+                ),
+                html.Div(
+                    id=self.uuid("paralell-coor-content"),
```
Comment (oysteoh) [inline ertviz/plugins/_parameter_comparison.py:32]: paralell-coor-content -> parallel-coor-content

**Comment by ManInFez** [comment]
This looks really nice! I have a few things I would change:
1. Add place holder text for the regex input box
2. Put "Search string" on same line as input box
3. Remove "ParCoor", not necessary and takes up alot of space
4. After pressing enter in the input box, selecting all visible would be cool :) 
5. Clicking on an option should default to append to already selected. Otherwise easy to loose the previously selected.  
6. Add "clear selection" for both plot and options widget

**Comment by xjules** [comment]
> This looks really nice! I have a few things I would change:

Thanks! :)

> 1. Add place holder text for the regex input box

Well, it's no really a regex (only a sub string currently). Shall we do it as regex though?

> 2. Put "Search string" on same line as input box

It was in the first version, no idea why I removed it ...

> 3. Remove "ParCoor", not necessary and takes up alot of space

I've left it there as a place holder if some checkboxes arrive at some point, but can go away.

> 4. After pressing enter in the input box, selecting all visible would be cool :)

Might be cool. It's not really a default behavior for the user  though; we can try it and left it as hint / comment.

> 5. Clicking on an option should default to append to already selected. Otherwise easy to loose the previously selected.
> 6. Add "clear selection" for both plot and options widget

I was thinking about two areas, one editable (to select from) and the other containing selected, but loose this idea as parameters are anyway visible as axes labels. Do you think about this as a default selection behavior, ie. only adding and one `clear all` button?

**Comment by oysteoh** [comment]
I think also we should make the search box case-insensitive? At least i like that - but I understand if there is other opinions as well - i just think it offers better usability 😄

**Comment by oysteoh** [comment]
And then i agree to what @ManInFez  says. And yes - i think it would be very useful with regexp - especially when there is a lot of params it would be really nice and offer great functionality !

Great job btw! 🚀

**Comment by ManInFez** [comment]
I do not think we need full regex, but at least having wild cards, like the dcc.Dropdown uses space as wildcards :)

**Comment by xjules** [comment]
> > 1. Add place holder text for the regex input box
> 
> Well, it's no really a regex (only a sub string currently). Shall we do it as regex though?

Done search syntax `"S1 S2 S3"` will get all parameters of type `aslkS1adlskjS2_eiuS3`
> 
> > 1. Put "Search string" on same line as input box
> 
> It was in the first version, no idea why I removed it ...

Done
> 
> > 1. Remove "ParCoor", not necessary and takes up alot of space
> 
> I've left it there as a place holder if some checkboxes arrive at some point, but can go away.
> 

Done
> > 1. After pressing enter in the input box, selecting all visible would be cool :)
> 
> Might be cool. It's not really a default behavior for the user though; we can try it and left it as hint / comment.
> 

Done, although need to fix the behavior of what's being selected and what's remained. 
> > 1. Clicking on an option should default to append to already selected. Otherwise easy to loose the previously selected.
> > 2. Add "clear selection" for both plot and options widget
> 

Done

File: ertviz/controllers/parameter_selector_controller.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            Input(parent.uuid("parameter-deactivator"), "value"),
+        ],
+    )
+    def update_parameters_options(selected_ensembles, filter_search, selected):
+        if not selected_ensembles:
+            raise PreventUpdate
+        selected = [] if not selected else selected
+
+        ensemble_id, _ = selected_ensembles.popitem()
+        ensemble = load_ensemble(parent, ensemble_id)
```
Comment (autt) [inline ertviz/controllers/parameter_selector_controller.py:38]: Are parameter options equal for all ensembles, or do we want to make sure that all selected ensembles are considered when finding valid parameter options? 🤔 See #111 #108 #109

Comment (xjules) [reply ertviz/controllers/parameter_selector_controller.py:38]: Good comment! Will need to make it as an intersection of ensemble parameters.

File: ertviz/controllers/parameter_selector_controller.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
+import regex
+import dash
+from dash.exceptions import PreventUpdate
+from dash.dependencies import Input, Output, State
+from ertviz.models import (
+    load_ensemble,
+)
+
+old_submit = 0
```
Comment (ManInFez) [inline ertviz/controllers/parameter_selector_controller.py:9]: this is not in use anymore

File: ertviz/controllers/parameter_selector_controller.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
+import regex
+import dash
+from dash.exceptions import PreventUpdate
+from dash.dependencies import Input, Output, State
+from ertviz.models import (
+    load_ensemble,
+)
+
+
+def filter_match(filter, key):
```
Comment (ManInFez) [inline ertviz/controllers/parameter_selector_controller.py:10]: `_filter_match`

File: ertviz/controllers/parameter_selector_controller.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+            Output(parent.uuid("parameter-deactivator"), "value"),
+        ],
+        [
+            Input(parent.uuid("parameter-selection-store"), "modified_timestamp"),
+        ],
+        [
+            State(parent.uuid("parameter-selection-store"), "data"),
+        ],
+    )
+    def update_parameter_selection(ts, shown_parameters):
```
Comment (ManInFez) [inline ertviz/controllers/parameter_selector_controller.py:116]: `ts` -> `_`

File: ertviz/models/plot_model.py [hunk 9]
```diff
@@ -265,6 +266,61 @@ def repr(self):
         return fig
 
 
+class ParallelCoordinates:
```
Comment (ManInFez) [inline ertviz/models/plot_model.py:269]: `ParallelCoordinatesPlotModel`

**Comment by ManInFez** [comment]
I think this looks great. Just some minor comments about naming style, but all in all very nice :)

**Review by ManInFez — APPROVED** [review]
Remeber to rebase and squash all non essential commits :)

**Pull request merged**